### PR TITLE
ControllerStage should complete when parsing of request failed and bed request response sent.

### LINKED ISSUE #20392: akka-http: postStop not called in graph stages for failed connections
The `postStop` handler of graph stages are not called when a connection reads an invalid http request. However, `preStart` is called.

Example proof of concept:

```scala
import java.util.concurrent.atomic.AtomicLong

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.http.scaladsl.server.Directives
import akka.stream._
import akka.stream.scaladsl.Flow
import akka.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}

object ExampleApp extends App with Directives {
  implicit val system = ActorSystem()
  implicit val mat = ActorMaterializer()

  val counter = new AtomicLong(0)

  val stage: GraphStage[FlowShape[HttpRequest, HttpResponse]] = new GraphStage[FlowShape[HttpRequest, HttpResponse]] {
    val in = Inlet[HttpRequest]("request.in")
    val out = Outlet[HttpResponse]("response.out")

    override def shape: FlowShape[HttpRequest, HttpResponse] = FlowShape.of(in, out)

    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
      override def preStart(): Unit = {
        counter.incrementAndGet()
      }

      override def postStop(): Unit = {
        counter.decrementAndGet()
      }

      setHandler(in, new InHandler {
        override def onPush(): Unit = {
          push(out, HttpResponse(entity = counter.get().toString))
        }
      })

      setHandler(out, new OutHandler {
        override def onPull(): Unit = pull(in)
      })
    }

  }
  val stageGraph = Flow.fromGraph(stage)

  Http().bindAndHandle(stageGraph, "0.0.0.0", 1234)
}
```

Here, one would assume that the response would always be the number `1` (assuming that it's called serially and so on.)

```
$ curl localhost:1234
1
```

However, for every request / connection that fails:

```
$ echo HELLO | nc localhost 1234
HTTP/1.1 400 Bad Request
Server: akka-http/2.4.4
Date: Mon, 25 Apr 2016 22:27:55 GMT
Connection: close
Content-Type: text/plain; charset=UTF-8
Content-Length: 26

Illegal HTTP message start
```

The counter is incremented (in `preStart`), but not decremented (which it would have been if `postStop` had been called).

Output after the above invalid http request:

```
$ curl localhost:1234
2
```

This may introduce memory leaks and/or invalid assumption of state (given the hopefully reasonable assumption that the calls to `preStart` and `postStop` should be equal).

### PR DESCRIPTION
**Pull request by rbudzko**
Might fix #20392 and/or #20391.

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Would you be able to make https://github.com/akka/akka/issues/20392 a reproducer test in ClientServerSpec?

**Comment by drewhk** [comment]
Looks correct, the only issue could be the concurrent propagation of cancellation and completion on the input and output side, but cancellation should be fine at the TCP level right?

**Comment by rbudzko** [comment]
I've verified cancellation by reading implementations wired in server blueprint, but please verify it as well. I'll work on regression tests when... feasible :)

**Comment by rbudzko** [comment]
@ktoso I hope this is what u meant.

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 1]
```diff
@@ -344,6 +346,53 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
...
+          val in = Inlet[HttpRequest]("request.in")
+          val out = Outlet[HttpResponse]("response.out")
+
+          override def shape: FlowShape[HttpRequest, HttpResponse] = FlowShape.of(in, out)
+
+          override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler with OutHandler {
+            override def preStart(): Unit = counter.incrementAndGet()
+            override def postStop(): Unit = counter.decrementAndGet()
+            override def onPush(): Unit = push(out, HttpResponse(entity = counter.get().toString))
+            override def onPull(): Unit = pull(in)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: maybe overwrite onUpstreamCompleted as well and count there in addition to the existing counter here?

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 2]
```diff
@@ -344,6 +346,53 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
...
+        def performFaultyRequest() = {
+          val socket = new Socket(hostname, port)
+          val os = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream, "UTF8"))
+
+          os.write("YOLO")
+          os.close()
+
+          socket.close()
+
+          Thread.sleep(1000)
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: remove the sleep

Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: and then use http://www.artima.com/docs-scalatest-2.0.M5/org/scalatest/concurrent/Eventually.html to check the shouldEqual 0

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala [hunk 3]
```diff
@@ -392,7 +441,7 @@ class ClientServerSpec extends WordSpec with Matchers with BeforeAndAfterAll wit
         val serverInSub = serverIn.expectSubscription()
         serverInSub.request(1)
         private val HttpRequest(POST, uri, List(Accept(Seq(MediaRanges.`*/*`)), Host(_, _), `User-Agent`(_)),
-          Chunked(`chunkedContentType`, chunkStream), HttpProtocols.`HTTP/1.1`) = serverIn.expectNext()
+        Chunked(`chunkedContentType`, chunkStream), HttpProtocols.`HTTP/1.1`) = serverIn.expectNext()
```
Comment (ktoso) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: wrong reformatting btw, should be `  `

**Comment by ktoso** [comment]
Yeah, almost there, let's tweak the test a little bit and LGTM

**Comment by rbudzko** [comment]
Can I download formatting configuration from somewhere?

**Comment by ktoso** [comment]
compile didn't reformat?

**Comment by ktoso** [comment]
// If it didn't we're working on fixing that https://github.com/akka/akka/pull/20591

**Comment by rbudzko** [comment]
It makes sense now :-)

File: akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala (hunk 1 above)
Comment (rbudzko) [inline akka-http-core/src/test/scala/akka/http/scaladsl/ClientServerSpec.scala]: @ktoso Did you meant onUpstreamFinish to calculate if upstream was closed for every request sent?

**Comment by rbudzko** [comment]
I think compile doesn't reformat for me in the end :-(

**Comment by drewhk** [comment]
@ktoso this looks LGTM to me may I merge?

**Comment by ktoso** [comment]
Yes, we're good here, thanks for the ping @drewhk !
LGTM

**Pull request merged**